src/test/protocol/TER_test.cpp

```diff
@@ -49,13 +52,214 @@ struct TER_test : public beast::unit_test::suite
         }
     }
 
+    // Helper template that makes sure two types are not convertible or
+    // assignable if not the same.
+    // o I1 one tuple index.
+    // o I2 other tuple index.
+    // o Tup is expected to be a tuple.
+    // It's a functor, rather than a function template, since a class template
+    // can be a template argument without being full specified.
+    template<std::size_t I1, std::size_t I2>
+    class NotConvertible
+    {
+    public:
+        template<typename Tup>
+        void operator()(Tup const& tup, beast::unit_test::suite&) const
+        {
+            // Entries in the tuple should not be convertible or assignable
+            // unless they are the same types.
+            using To_t = std::decay_t<decltype (std::get<I1>(tup))>;
+            using From_t = std::decay_t<decltype (std::get<I2>(tup))>;
+            static_assert (!std::is_same<From_t, To_t>::value ^
+                std::is_convertible<From_t, To_t>::value, "Convert err");
+            static_assert (!std::is_same<To_t, From_t>::value ^
+                std::is_constructible<To_t, From_t>::value, "Construct err");
+            static_assert (!std::is_same <To_t, From_t>::value ^
+                std::is_assignable<To_t&, From_t const&>::value, "Assign err");
+
+            // Assignment or conversion from integer to type should never work.
+            static_assert (
+                ! std::is_convertible<int, To_t>::value, "Convert err");
+            static_assert (
+                ! std::is_constructible<To_t, int>::value, "Construct err");
+            static_assert (
+                ! std::is_assignable<To_t&, int const&>::value, "Assign err");
```

Comment:
I think just `std::is_assignable<To_t&, int>`. Why the `int const&`?

Reply:
Mostly a reaction to uses of `is_assignable` being tricky to get right (as I learned the hard way), since it defaults types to `rvalue`s.  The From argument being a `const &` is the more typical case, since that's the signature of a defaulted copy assignment operator.  But, it's true, either would probably work.